### PR TITLE
tbb GausHitFinder_module

### PR DESCRIPTION
**Pull request by cerati**
This pull request changes the loop over wires in the GausHitFinder_module from a simple for loop to a tbb::parallel_for. Concurrent pushing of results in the output vector is handled via tbb::concurrent_vector. Credits to @sberkman for the initial implementation. Thread scaling performance is currently under study. The PR also includes a bug fix in the PeakFitterMrqdt_tool, so that the number of degrees of freedom (NDF) is now properly returned.

### CODE REVIEW

File: larreco/HitFinder/GausHitFinder_module.cc [hunk 1]
```diff
@@ -53,6 +53,10 @@
 #include "TH1F.h"
 #include "TMath.h"
 
+#include "tbb/tbb.h"
+#include "tbb/blocked_range.h"
```
Comment (knoepfel) [inline larreco/HitFinder/GausHitFinder_module.cc]: I suspect you can remove the `tbb/blocked_range` dependency.

File: larreco/HitFinder/GausHitFinder_module.cc [hunk 2]
```diff
@@ -454,12 +476,17 @@ void GausHitFinder::produce(art::Event& evt)
...
 
                     const recob::Hit hit(hitcreator.move());
 
+		    hitstruct hitstruct_local;
+		    hitstruct_local.hit_tbb=std::move(hit);
+		    hitstruct_local.wire_tbb=wire;
+
                     // This loop will store ALL hits
-                    allHitCol.emplace_back(std::move(hit), wire);
+		    hitstruct_vec.push_back(hitstruct_local);
```
Comment (knoepfel) [inline larreco/HitFinder/GausHitFinder_module.cc]: Recommend more concise/efficient pattern:
```.cpp
hitstruct tmp{std::move(hit), wire};
hitstruct_vec.push_back(std::move(tmp));
```

File: larreco/HitFinder/GausHitFinder_module.cc [hunk 3]
```diff
@@ -499,18 +526,29 @@ void GausHitFinder::produce(art::Event& evt)
...
                     // Copy the hits we want to keep to the filtered hit collection
                     for(const auto& filteredHit : filteredHitVec)
-                        if (!fHitFilterAlg || fHitFilterAlg->IsGoodHit(filteredHit))
-                            filteredHitCol->emplace_back(filteredHit, wire);
-                }
+		      if (!fHitFilterAlg || fHitFilterAlg->IsGoodHit(filteredHit)){
+			hitstruct filthitstruct_local;
+			filthitstruct_local.hit_tbb=filteredHit;
+			filthitstruct_local.wire_tbb=wire;
+			filthitstruct_vec.push_back(filthitstruct_local);
```
Comment (knoepfel) [inline larreco/HitFinder/GausHitFinder_module.cc]: Recommend more concise/efficient pattern:
```.cpp
hitstruct tmp{std::move(filteredHit), wire};
filthitstruct_vec.push_back(std::move(tmp));
```

**Review by knoepfel — CHANGES_REQUESTED** [review]
Relatively minor changes requested.  Looks good.

**Comment by cerati** [comment]
@knoepfel's suggestions above are implemented in the last commit (thanks!)

**Review by knoepfel — APPROVED** [review]

**Comment by knoepfel** [comment]
@cerati, looks good!

**Comment by knoepfel** [comment]
approve

**Comment by knoepfel** [comment]
trigger build

**Comment by lgarren** [comment]
trigger build

Previous CI tests did not include uboonecode, which is now updated.  Apparently sbndcode still lags behind.

**Comment by lgarren** [comment]
trigger build

Previous CI tests did not include uboonecode, which is now updated.  Apparently sbndcode still lags behind.

**Pull request merged**